src/plots/cartesian/dragbox.js

```diff
@@ -578,31 +705,44 @@ module.exports = function dragBox(gd, plotinfo, x, y, w, h, ns, ew) {
                 editY = editY && isInY;
             }
 
-            var xScaleFactor = editX ? xa2._length / viewBox[2] : 1,
-                yScaleFactor = editY ? ya2._length / viewBox[3] : 1;
+            var xScaleFactor = editX ? viewBox[2] / xa2._length : 1,
+                yScaleFactor = editY ? viewBox[3] / ya2._length : 1;
 
             var clipDx = editX ? viewBox[0] : 0,
                 clipDy = editY ? viewBox[1] : 0;
 
-            var fracDx = editX ? (viewBox[0] / viewBox[2] * xa2._length) : 0,
-                fracDy = editY ? (viewBox[1] / viewBox[3] * ya2._length) : 0;
+            // modify these if needed for linked axes
+            if(editX && !editY && xaLinked.indexOf(ya2) !== -1) {
+                yScaleFactor = xScaleFactor;
+                clipDy = ya2._length * (1 - yScaleFactor) / 2;
+
+                // update range for the linked axis.
+                ya2.range = ya2._r.slice();
+                scaleZoom(ya2, yScaleFactor);
+            }
+            if(editY && !editX && (isSubplotConstrained ? xaLinked : yaLinked).indexOf(xa2) !== -1) {
+                xScaleFactor = yScaleFactor;
+                clipDx = xa2._length * (1 - xScaleFactor) / 2;
+                xa2.range = xa2._r.slice();
+                scaleZoom(xa2, xScaleFactor);
+            }
 
-            var plotDx = xa2._offset - fracDx,
-                plotDy = ya2._offset - fracDy;
+            var plotDx = xa2._offset - clipDx / xScaleFactor,
+                plotDy = ya2._offset - clipDy / yScaleFactor;
 
             fullLayout._defs.selectAll('#' + subplot.clipId)
                 .call(Drawing.setTranslate, clipDx, clipDy)
-                .call(Drawing.setScale, 1 / xScaleFactor, 1 / yScaleFactor);
+                .call(Drawing.setScale, xScaleFactor, yScaleFactor);
 
             subplot.plot
                 .call(Drawing.setTranslate, plotDx, plotDy)
-                .call(Drawing.setScale, xScaleFactor, yScaleFactor)
+                .call(Drawing.setScale, 1 / xScaleFactor, 1 / yScaleFactor)
 
                 // This is specifically directed at scatter traces, applying an inverse
                 // scale to individual points to counteract the scale of the trace
                 // as a whole:
                 .select('.scatterlayer').selectAll('.points').selectAll('.point')
-                    .call(Drawing.setPointGroupScale, 1 / xScaleFactor, 1 / yScaleFactor);
+                    .call(Drawing.setPointGroupScale, xScaleFactor, yScaleFactor);
```

Comment:
no, I just flipped the definitions of x/yScaleFactor since I noticed they were being inverted in most of their uses.